hazelcast/src/test/java/com/hazelcast/test/compatibility/SamplingSerializationService.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.test.compatibility;
+
+import com.hazelcast.core.ManagedContext;
+import com.hazelcast.core.PartitioningStrategy;
+import com.hazelcast.internal.serialization.InternalSerializationService;
+import com.hazelcast.internal.serialization.PortableContext;
+import com.hazelcast.nio.BufferObjectDataInput;
+import com.hazelcast.nio.BufferObjectDataOutput;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.nio.serialization.PortableReader;
+import com.hazelcast.test.TestEnvironment;
+
+import java.io.IOException;
+import java.nio.ByteOrder;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+/**
+ * Serialization service that intercepts and samples serialized objects.
+ * Employed to gather samples of serialized objects used in member-to-member communication during test suite
+ * execution with mock network.
+ *
+ * @see TestEnvironment#SAMPLE_SERIALIZED_OBJECTS
+ */
+public class SamplingSerializationService implements InternalSerializationService {
+
+    static final ConcurrentMap<String, List<byte[]>> SERIALIZED_SAMPLES_PER_CLASS_NAME =
+            new ConcurrentHashMap<String, List<byte[]>>(1000);
+    private static final int MAX_SERIALIZED_SAMPLES_PER_CLASS = 5;
+    // utility strings to locate commonly used test classes without "Test" / ".test" in their class name
+    private static final String SAMPLE_OBJECTS_TEST_CLASS_NAME = "com.hazelcast.query.SampleObjects";
+    private static final String COMPLEX_DATA_STRUCTURE_TEST_CLASS_NAME =
+            "com.hazelcast.query.impl.extractor.specification.ComplexDataStructure";
+    private static final String TEST_CLASS_SUFFIX = "Test";
+    private static final String TEST_PACKAGE_INFIX = "Test";
+
+    protected final InternalSerializationService delegate;
+
+    public SamplingSerializationService(InternalSerializationService delegate) {
+        this.delegate = delegate;
+    }
+
+    // record the given object, then return it
+    protected static <T> T sampleObject(T obj, byte[] serializedObject) {
```

Comment:
I would move this method down to the other static sampling methods, then they are grouped together.